drivers/led/ncp5623c.c

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) 2023 NXP Semiconductors
+ * Copyright (c) 2023 Cognipilot Foundation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT onnn_ncp5623c
+
+/**
+ * @file
+ * @brief NCP5623C LED driver
+ *
+ * The NCP5623C is a 3-channel LED driver that communicates over I2C.
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/led.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(ncp5623c, CONFIG_LED_LOG_LEVEL);
+
+#include "led_context.h"
+
+#define NCP5623_LED_CURRENT 0x20
+#define NCP5623_LED_PWM0    0x40
+#define NCP5623_LED_PWM1    0x60
+#define NCP5623_LED_PWM2    0x80
+
+#define NCP5623C_CHANNEL_COUNT 3
+
+/* Brightness limits */
+#define NCP5623C_MIN_BRIGHTNESS 0
+#define NCP5623C_MAX_BRIGHTNESS 0x1f
+
+static const uint8_t led_channels[] = {NCP5623_LED_PWM0, NCP5623_LED_PWM1, NCP5623_LED_PWM2};
+
+struct ncp5623c_config {
+	struct i2c_dt_spec bus;
+	uint8_t num_leds;
+	const struct led_info *leds_info;
+};
+
+struct ncp5623c_data {
+	struct led_data dev_data;
+};
+
+static const struct led_info *ncp5623c_led_to_info(const struct ncp5623c_config *config,
+						   uint32_t led)
+{
+	if (led < config->num_leds) {
+		return &config->leds_info[led];
+	}
+
+	return NULL;
+}
+
+static int ncp5623c_get_info(const struct device *dev, uint32_t led, const struct led_info **info)
+{
+	const struct ncp5623c_config *config = dev->config;
+	const struct led_info *led_info = ncp5623c_led_to_info(config, led);
+
+	if (!led_info) {
+		return -EINVAL;
+	}
+
+	*info = led_info;
+
+	return 0;
+}
+
+static int ncp5623c_set_color(const struct device *dev, uint32_t led, uint8_t num_colors,
+			      const uint8_t *color)
+{
+	const struct ncp5623c_config *config = dev->config;
+	const struct led_info *led_info = ncp5623c_led_to_info(config, led);
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (num_colors != 3) {
+		LOG_ERR("%s: invalid number of colors: got=%d, expected=%d", dev->name, num_colors,
+			3);
+		return -ENOSUP;
+	}
+
+	uint8_t buf[6] = {0x70, NCP5623_LED_PWM0, 0x70, NCP5623_LED_PWM1, 0x70, NCP5623_LED_PWM2};
+
+	buf[1] = buf[1] | color[0] / 8;
+	buf[3] = buf[3] | color[1] / 8;
+	buf[5] = buf[5] | color[2] / 8;
+
+	return i2c_burst_write_dt(&config->bus, NCP5623_LED_CURRENT | NCP5623C_MAX_BRIGHTNESS, buf,
+				  sizeof(buf));
+}
+
+static int ncp5623c_set_brightness(const struct device *dev, uint32_t led, uint8_t value)
+{
+	const struct ncp5623c_config *config = dev->config;
+	const struct led_info *led_info = ncp5623c_led_to_info(config, led);
+	int ret = 0;
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (value > 100) {
+		return -EINVAL;
+	}
+
+	if (led_info->num_colors == 3) {
+		uint8_t colors[3];
+
+		colors[0] = value * 255 / 100;
+		colors[1] = colors[0];
+		colors[2] = colors[0];
```

Comment:
This doesn't work. You must keep the color ratio when applying the new brightness.

You could use something like:
```
uint8_t colors[3];
uint8_t max_color = 0;

/* Read colors from PWM */

value = value * NCP5623C_MAX_BRIGHTNESS / 100;

for (i = 0; i < num_colors; i++) {
        max_color = MAX(max_color, colors[i]);
}
for (i = 0; i < num_colors; i++) {
        colors[i] = colors[i] * value / max_color;
}

/* Write colors to PWM */
```

But is probably better to let it unsupported (`return -ENOSUP;`). Because after all it is not supported by the controller...